Dockerfile

```diff
@@ -1,46 +1,60 @@
-RUN dpkg --add-architecture i386 && \
-            apt-get update && \
-            apt-get install -y python-pip cython vim build-essential ccache git gcc openjdk-8-jdk \
-                lsb-release unzip wget curl python-dev zlib1g-dev ant xsel xclip \
-                zlib1g:i386 libncurses5:i386 libstdc++6:i386 autoconf automake libtool libffi-dev libltdl-dev && \
-            apt-get clean && \
-            pip install pip --upgrade && \
-            pip install cython buildozer
-
-# Create an unprivileged user and run as that user, to please Buildozer
-RUN adduser -u 1000 kivy && \
-            mkdir -p /data /home/kivy/.buildozer && chown 1000 /data && \
-            chown -R kivy.kivy /home/kivy/.buildozer
-USER kivy
 WORKDIR /home/kivy
 
-# Copy in the Buildozer spec file, and update build system (download NDK/SDK, build Python, etc)
-COPY buildozer.spec .
-RUN buildozer android update
-
-# Because COPY below won't run as kivy user, make sure it will then have permission to unzip to ./code
-USER root
-RUN mkdir -p /home/kivy/code && chown kivy.kivy /home/kivy/code
-USER kivy
-
-# Copy in the code and resources
-COPY code ./code
-COPY kolibri*.whl .
-RUN unzip "kolibri*.whl" "kolibri/*" -d code/
-COPY resources ./resources
-
-# Remove the default loading page, so that it will be replaced with our own version
-RUN rm .buildozer/android/platform/build/dists/kolibri/webview_includes/_load.html
-
-# Build the APK
-RUN buildozer android debug
-
-# Copy the generated APK up the root for easy discovery and extraction
-USER root
-RUN cp /home/kivy/bin/*.apk /
+# Install the dependencies for the build system
+RUN dpkg --add-architecture i386 && \
+    apt-get update && apt-get install -y \
+    ant \
+    autoconf \
+    automake \
+    build-essential \
+    ccache \
+    curl \
+    cython \
+    gcc \
+    git \
+    libffi-dev \
+    libltdl-dev\
+    libncurses5:i386 \
+    libstdc++6:i386 \
+    libtool \
+    lsb-release \
+    openjdk-8-jdk \
+    python-dev \
+    python-pip \
+    unzip \
+    vim \
+    wget \
+    xclip \
+    xsel \
+    zlib1g-dev \
+    zlib1g:i386 \
+    && apt-get clean && \
+    pip install pip --upgrade && \
+    pip install cython buildozer
+
+# would be nice if we could batch this into the bash file
+# Copy over files vital to build environment
+# NOTE: Technically, the only file necessary here is buildozer.spec
+COPY Makefile buildozer.spec ./
+
+# Set up build environment
+RUN useradd -l kivy && \
+  chown -R kivy:kivy /home/kivy &&\
+  su kivy -c make updatedependencies
+
+# Keeping these copies separate to keep cache valid as long as possible in setup
+# for kolibri builds. Buildozer takes a while to rebuild itself
+# Copy over kolibri-apk specific build files
+COPY ./src  src/
+
+# Extract .whl files and build the apk
+RUN chown -R kivy:kivy /home/kivy && ls -la;\
+  su kivy -c "\
+    make removeloadingpage && \
+    make extractkolibriwhl && \
+    make generateversion && \
+    make builddebugapk"
```

Comment:
The kicker (and the main reason builds are still going to be pretty long) is that buildozer doesn't check for android ndk and other packages until we build the apk here. Their documentation assures that this will only happen the first time you run the command but because we're building a new image every build, every time is the first time.

Reply:
Hmm. Now during the docker build, the packages are downloaded in `builddebugapk` not in `updatedependencies` which is causing the default loading page to be re added.